main/src/ca/uwaterloo/flix/language/phase/optimizer/RecursionRewriter.scala

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright 2025 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.optimizer
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.MonoAst.Expr
+import ca.uwaterloo.flix.language.ast.shared.{BoundBy, Scope}
+import ca.uwaterloo.flix.language.ast.{MonoAst, SourceLocation, Symbol, Type}
+import ca.uwaterloo.flix.language.dbg.AstPrinter.DebugMonoAst
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import java.util.concurrent.atomic.AtomicBoolean
+import scala.collection.mutable
+
+/**
+  * Rewrites functions that recursively call themselves in tail-position to
+  * non-recursive functions with a recursive local def.
+  */
+object RecursionRewriter {
+
+  def run(root: MonoAst.Root)(implicit flix: Flix): MonoAst.Root = flix.phase("RecursionRewriter") {
+    val newDefs = ParOps.parMapValues(root.defs)(visitDef)
+    root.copy(defs = newDefs)
+  }
+
+  private def visitDef(defn: MonoAst.Def)(implicit flix: Flix): MonoAst.Def = {
+    // 1. Check that every recursive call is in tail position
+    //    Return a set of alive parameters, i.e, function parameters that are changed in a recursive call (if it is an Expr.Var with the same symbol, then it is dead).
+    implicit val ctx: LocalContext = LocalContext.mk()
+    val allRecursiveCallsInTailPos = checkTailPosition(defn.exp, tailPos = true)(defn.sym, defn.spec.fparams, ctx, flix)
+    val containsRecursiveCall = ctx.isRecursive.get()
```

Comment:
The `isRecursive` predicate in `ctx` could probably be derived from collecting all the self-`ApplyDef` exprs. If that list is non-empty, then at least one recursive call exists.